ducktape/tests/runner.py

```diff
@@ -156,30 +221,40 @@ def _handle_ready(self, event):
         self.receiver.send(
                 self.event_response.ready(event, self.session_context, test_context, subcluster))
 
-        # Test is considered "active" as soon as we receive the ready request
-        self.active_tests[event["test_id"]] = event
-
     def _handle_log(self, event):
         self.receiver.send(self.event_response.log(event))
         self._log(event["log_level"], event["message"])
 
     def _handle_finished(self, event):
+        test_id = event["test_id"]
         self.receiver.send(self.event_response.finished(event))
 
+        result = event['result']
+        if result.test_status == FAIL and self.exit_first:
+            self.stop_testing = True
+
         # Transition this test from running to finished
-        test_id = event["test_id"]
         del self.active_tests[test_id]
         self.finished_tests[test_id] = event
-        self.results.append(event['result'])
+        self.results.append(result)
 
         # Free nodes used by the test
         subcluster = self._test_cluster[test_id]
         self.cluster.free(subcluster.alloc(len(subcluster)))
         del self._test_cluster[test_id]
 
-        if len(self.staged_tests) + len(self.active_tests) > 0 and self.session_context.max_parallel == 1:
+        # Join on the finished test process
+        self._client_procs[test_id].join()
+
+        if self._should_print_delimiter:
             terminal_width, y = get_terminal_size()
-            print "~" * int(2 * terminal_width / 3)
+            self._log(logging.INFO, "~" * int(2 * terminal_width / 3))
+
+    @property
+    def _should_print_delimiter(self):
+        return self.session_context.max_parallel == 1 and \
+            not self.stop_testing and \
```

Comment:
Do you really want to filter on `stop_testing`? Won't that just sort of cut off some of the output for a test that's in progress? In fact, is it even relevant if you've already checked for `max_parallel == 1` since that condition guarantees there's only 1 test running at a time and if you marked `stop_testing` that must mean it already failed?